src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -168,28 +168,108 @@ internal bool TryAddWithoutValidation(HeaderDescriptor descriptor, string? value
-                if (enumerator.MoveNext())
+                return true;
+            }
+
+            // read the store of header values
+            // The header store contain a single raw string value when header values are single.
+            // The header store contain HeaderStoreItemInfo that wraps around a List<string> with header values when they are more than once
+
+            ref object? storeValueRef = ref GetValueRefOrAddDefault(descriptor);
+            object? storeValue = storeValueRef;
+
+            // check whether the header store contains already a value
+            // or header values count is more than one
+            // => allocate a HeaderStoreItemInfo on store that wraps around a List<string> of header values
+            if (storeValue is not null || (count.HasValue && count.Value > 1))
+            {
+                if (storeValue is not HeaderStoreItemInfo info)
+                {
+                    storeValueRef = info = new HeaderStoreItemInfo { RawValue = storeValue };
+                }
+
+                object? rawValue = info.RawValue;
+                if (rawValue is not List<string> rawValues)
                 {
-                    ref object? storeValueRef = ref GetValueRefOrAddDefault(descriptor);
-                    Debug.Assert(storeValueRef is not null);
+                    info.RawValue = rawValues = new List<string>();
 
-                    object value = storeValueRef;
-                    if (value is not HeaderStoreItemInfo info)
+                    if (rawValue != null)
                     {
-                        Debug.Assert(value is string);
-                        storeValueRef = info = new HeaderStoreItemInfo { RawValue = value };
+                        if (count.HasValue)
+                        {
+                            rawValues.EnsureCapacity(count.Value + 1);
+                        }
+
+                        rawValues.Add((string)rawValue);
                     }
+                }
+
+                if (count.HasValue)
+                {
+                    rawValues.EnsureCapacity(count.Value);
+                }
 
-                    do
+                if (valuesArray != null)
+                {
+                    for (int i = 0; i < valuesArray.Length; i++)
                     {
-                        AddRawValue(info, enumerator.Current ?? string.Empty);
+                        rawValues.Add(valuesArray[i] ?? string.Empty);
+                    }
+                }
+                else if (valuesList != null)
+                {
+                    for (int i = 0; i < valuesList!.Count; i++)
+                    {
+                        rawValues.Add(valuesList[i] ?? string.Empty);
+                    }
+                }
+                else
+                {
+                    foreach (string? value in values)
+                    {
+                        rawValues.Add(value ?? string.Empty);
+                    }
+                }
```

Comment:
The `else` block below is already handling the case where the initial `storeValue` was null. I don't think we need to handle the `IEnumerable` case here.